hir_typeck: be more conservative in making "note caller chooses ty param" note

jieyouxu opened the pull request:
In https://github.com/rust-lang/rust/pull/122195 I added a "caller chooses ty for type param" note for when the return expression type a.k.a. found type does not match the expected return type.

#126547 found that this note was confusing when the found return type *contains* the expected type, e.g.

```rs
fn f<T>(t: &T) -> T {
    t
}
```

because the found return type `&T` will *always* be different from the expected return type `T`, so the note was needlessly redundant and confusing.

This PR addresses that by not making the note if the found return type contains the expected return type.

r? @fmease (since you reviewed the original PR)

Fixes https://github.com/rust-lang/rust/issues/126547

jieyouxu committed:
tests: update tests for more conservative return ty mismatch note

fmease commented on compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:
```diff
@@ -874,9 +874,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     // are not, the expectation must have been caused by something else.
-                    debug!("return type {:?}", hir_ty);
+                    debug!("return type (hir::Ty) {:?}", hir_ty);
                     let ty = self.lowerer().lower_ty(hir_ty);
-                    debug!("return type {:?}", ty);
+                    debug!("return type (ty) {:?}", ty);
                     debug!("expected type {:?}", expected);
```
Btw, `tracing` has a nice shorthand for this but it doesn't matter (for future reference).

```rs
debug!(?hir_ty, "return type");
let ty = self.lowerer().lower_ty(hir_ty);
debug!(?ty, "return type {:?}");
debug!(ty = ?expected, "expected type");
```

which will print:

```
return type hir_ty=...
return type ty=...
expected type exp ty=...
```

jieyouxu replied:
(I actually did know that shorthand, I just copied over what existed lol)

jieyouxu replied:
... but now that you mentioned it, it's indeed nicer so I changed it ^^

fmease reviewed (approved):
Thanks!

fmease commented:
@bors r+ rollup

bors commented:
:pushpin: Commit a5eff62d4c90c790a6a1626311ab6d49c8b2eaaf has been approved by `fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fmease a5eff62d4c90c790a6a1626311ab6d49c8b2eaaf -->
<!-- homu: {"type":"Approved","sha":"a5eff62d4c90c790a6a1626311ab6d49c8b2eaaf","approver":"fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #126623) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jieyouxu committed:
hir_typeck: be more conservative in making "note caller chooses ty param" note

- Avoid "caller chooses ty for type param" note if the found type a.k.a.
  the return expression type *contains* the type parameter, because e.g.
  `&T` will always be different from `T` (i.e. "well duh").
- Rename `note_caller_chooses_ty_for_ty_param` to
  `try_note_caller_chooses_ty_for_ty_param` because the note is not
  always made.

Issue: https://github.com/rust-lang/rust/issues/126547

jieyouxu commented:
@rustbot review

fmease commented:
@bors r+

bors commented:
:pushpin: Commit 939026c8fba0d6b9edc2e2cdfbbf6dd6fb575662 has been approved by `fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fmease 939026c8fba0d6b9edc2e2cdfbbf6dd6fb575662 -->
<!-- homu: {"type":"Approved","sha":"939026c8fba0d6b9edc2e2cdfbbf6dd6fb575662","approver":"fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
